[Windows] [Installer] Wix bootstrapper bundle

sblaisot opened the pull request:
This PR is intended to fix [lp:1663380](https://bugs.launchpad.net/mixxx/+bug/1663380) (installation failing on windows when MS visual C++ 2015 redistributable is not already installed).

- [X] Remove vcredist from main installer
- [X] Add bootstrapper (bundle) to chain vcredist with Mixxx installer
- [x] Only bundle vcredist related to msi arch instead of all vcredist arch with each package
- [x] Enforce MS visual C++ redistributable detection in Mixxx MSI
- [x] Avoid restore point creation to speed up installation
- [x] fix uninstallation
- [x] UI polish
  - [x] redesign screens
  - [x] update images
  - [x] custom translatable texts
- [x] make bootstrapper localizable
- [x] Update installer translations
  - [x] push new text to transifex
  - [x] merge new  available translations from transifex
- [x] handle 64 bits installer on win32 (warning + exit)
- [x] Move "launch Mixxx after installation" from MSI to bootstrapper
- [x] Fix uninstallation and repair success messages
- [x] Make localization languages dynamic based on available files instead of hard-coded into bundle.wxs
- [x] Test, test and more tests

Please note that with this change, we will distribute exe installer again (but with an MSI bundled).

**ready to merge.**

sblaisot committed:
(wix) remove vcredist from the main msi package

sblaisot committed:
(wix) Add bootstrapper to handle vcredist properly

sblaisot commented:
Windows users tests required:
1. uninstall Mixxx
2. Uninstall all MS visual C++ 2015 redistributables (x86 and x64)
3. Download https://ci.appveyor.com/project/mixxxdj/mixxx/build/master-2100/job/skpt4a9wlv0fa12y/artifacts
4. try to install

please let me know if this works

cc/ @daschuer

sblaisot committed:
(wix) update bootstrapper sidebar image

sblaisot committed:
(wix) fix sidebar image

sblaisot committed:
(wix) bundle vcredist only for the selected arch

sblaisot committed:
(wix) disable restore point creation in bootstrapper to speed up installation

sblaisot committed:
(wix) update sidebar to proper size

Be-ing commented:
> Windows users tests required

Post to the forum with those instructions.

sblaisot commented:
it appears we have a bug in the uninstallation process. installation seems to work fine but you can not uninstall :(

sblaisot committed:
(wix) fix x64 redistributable

sblaisot commented:
> Post to the forum with those instructions.

I'll do that but we currently have major issues I need to fix before letting users try this. It's only suitable for advanced developpers right now ;)

sblaisot commented:
> it appears we have a bug in the uninstallation process. installation seems to work fine but you can not uninstall :(

Probably related to my computer only, windows package cache/registry seems broken by too much trying. Updating version number fixed the uninstall functionnality, no more bug.

sblaisot committed:
(wix) properly handle c++ redistributable repair and uninstall

sblaisot committed:
(wix) invert bundle redistributable bitwidth logic

sblaisot committed:
(wix) show version in bootstrapper

sblaisot committed:
(wix) define it's own UpgradeCode for bootstrapper to avoid collision with main Mixxx package

sblaisot committed:
(wix) fix indentation in bundle.wxs

sblaisot committed:
(wix) update installer translations from transifex

sblaisot committed:
(wix) Enforce MS Visual C++ Redistributable presence in Mixxx MSI

sblaisot committed:
(wix) redesign bootstrapper UI and make strings translatable

sblaisot committed:
(wix) update translations with new strings

sblaisot committed:
fix translations

sblaisot committed:
(wix) fix vcredist installation enforcement

sblaisot committed:
(wix) update vcredist friendly names

sblaisot committed:
(wix) prevent 64bits bundle installation on 32bits OS

esbrandt commented on build/wix/Localization/mixxx_en-us.wxl:
```diff
@@ -24,6 +24,7 @@
    <String Id="bitWidthMismatch32">You are about to install the 32-bit version but a 64-bit one is already installed. This can lead to undesired behavior. Please un-install the existing Mixxx first, or download the 64-bit installation package to upgrade (unless you know exactly what you are doing.)</String>
    <String Id="bitWidthMismatch64">You are about to install the 64-bit version but a 32-bit one is already installed. This can lead to undesired behavior. Please un-install the existing Mixxx first, or download the 32-bit installation package to upgrade (unless you know exactly what you are doing.)</String>
    <String Id="OlderVersionFound">An older version of Mixxx has been detected. As we changed installer, we can not smoothly upgrade. Please uninstall older version before installing this one.</String>
+   <String Id="VCredistNotFound">Installer was not able to find Microsoft Visual C++ 2015 Redistributable on this system. This is likely an installer bug that should be reported to Mixxx developpers.</String>
```
Typo: developpers -> developers

sblaisot replied:
Thanks, typo fixed.
I also updated translation source on transifex.

esbrandt replied:
> I also updated translation source on transifex.

Translating got a bit harder with the ampersands and variables.
In the past, we had conflicting shortcuts with some languages, because translators didn't fully understand what the ampersands do.

Unfortunately, the Windows installer can not share our extensive translation memory and glossary with the rest of the project (application, manual, website) because the base language is ``en_us`` instead of ``en``.

The installer got it right IMO, we need to carry over the base language for the other parts of the project to ``en_us``. Will check the requirements and submit a PR.

sblaisot committed:
(wix) variabilize bundle name and bundle manufacturer

sblaisot committed:
Revert "(wix) fix vcredist installation enforcement"

This reverts commit 5af1279858cb4609be9edc7c199bcbabb69a6c4e.

sblaisot committed:
(wix) fix typo

sblaisot committed:
(wix) fix vcredist enforcement for 32-bits windows

sblaisot committed:
(wix) fix vcredist installation detection in bootstrapper
Allow best detection so don't install if already installed and allow for repair and uninstall from within the bootstrapper

sblaisot committed:
(wix) update translations after fixing typo

sblaisot committed:
(wix) move installation directory selection and launch button at installation end to bootstrapper

sblaisot committed:
Merge branch 'wix-bundle' of https://github.com/sblaisot/mixxx into wix-bundle

sblaisot committed:
(wix) clean old test batch file and update README.md

sblaisot committed:
(wix) Update Readme.md with bootstrapper info

sblaisot committed:
(wix) fix README.md

sblaisot committed:
(wix) Fix uninstallation and repair success messages

sblaisot committed:
(wix) update translations after new strings has been added

sblaisot committed:
(wix) Make bundle localization dynamic

esbrandt commented on build/wix/Localization/mixxx_cs-CZ.wxl:
```diff
@@ -50,4 +51,48 @@
+  <String Id="HelpText">/install | /repair | /uninstall | /layout [directory] - installs, repairs, uninstalls or&#13;&#10;    creates a complete local copy of the bundle in directory.&#13;&#10;    Install is the default.&#13;&#10;&#13;&#10; /passive | /quiet -  displays minimal UI with no prompts or displays no UI&#13;&#10;    and no prompts. By default UI and all prompts are displayed.&#13;&#10; ...
+  <String Id="HelpCloseButton">&amp;Close</String>
+  <String Id="InstallHeader">Welcome</String>
+  <String Id="InstallMessage">Setup will install Mixxx on your computer.&#13;&#10;It will first install required third party libraries, then Mixxx itself.&#13;&#10;Click install to continue or Close to exit.</String>
```
Capitalization:
``Click install to continue or Close to exit``
 ->
``Click Install to continue or Close to exit``

sblaisot replied:
Thanks. Fixed.

sblaisot committed:
(wix) fix typo and update translations

esbrandt commented on build/wix/Localization/mixxx_en-us.wxl:
```diff
@@ -43,5 +44,65 @@
+   <String Id="ConfirmCancelMessage">Are you sure you want to cancel?</String>
+   <String Id="ExecuteUpgradeRelatedBundleMessage">Previous version</String>
+   <String Id="HelpHeader">Setup Help</String>
+   <String Id="HelpText">/install | /repair | /uninstall | /layout [directory] - installs, repairs, uninstalls or
```
*Nits found while translating:*
Spare space after ``-``
``/quiet -  ``
Spare space before ``-`` 
``/norestart   -``

I´d write in full ``UI`` -> ``user interface``

sblaisot replied:
Thanks, updated.
In fact these ones come from upstream ;)

sblaisot committed:
(wix) fix small typo/nits

sblaisot committed:
(wix) UI -> user interface

esbrandt commented:
Tested ``mixxx-2.1.0-alpha-pre-master-PR1306-git6322-release-x64`` with a Windows 7 SP1 VM, and it works quite well.

* Installs the redist package, then the application itself.
* Uninstall does not work. Setup says that the uninstall was successful, in fact no files/registry entries are removed. Will test with Win 10 VM again.

The launch screen has some potential minor issues:

* It says ``Mixxx`` 5 times (red circles). 
* The text is cropped with german translation active, potentially with other languages too ( green rectangle)
   What about ``Setup will first install required third party libraries, then Mixxx itself on your computer. Click Install to continue or Close to exit.``

* The launch image - Is nice, but potentially outdated with this or the next version ( e.g. change of default skin / skin layout). I´d prefer to not have the GUI depicted. What about just the icon?

![image 3](https://user-images.githubusercontent.com/4525897/28249983-15cf0f3a-6a60-11e7-8bf7-196412090ed3.png)

Be-ing commented:
> The launch image - Is nice, but potentially outdated with this or the next version ( e.g. change of default skin / skin layout). I´d prefer to not have the GUI depicted. What about just the icon?

I agree, just the Mixxx logo would work well.

esbrandt commented:
> Uninstall does not work. Setup says that the uninstall was successful, in fact no files/registry entries are removed. Will test with Win 10 VM again.

Same with Win 10 VM as with Win 7 VM.

sblaisot commented:
@esbrandt can you please send me the uninstall log you can find in %TEMP% ?
I need that to see why it doesn't unisntall properly.

also, you can try to extract the embedded MSI with
`mixxx-2.1.0-alpha-pre-master-PR1306-git6322-release-x64.exe -layout .`
(don't forget the dot at the end of command line to extract in the current working directory)
then uninstall using the MSI with 
`msiexec /x mixxx-2.1.0-alpha-pre-master-PR1306-git6322-release-x64.msi`
and report if it uninstalls fine

sblaisot commented:
ok, I will work on screen redisign, but please note that screen layout is highly customizable with bootstrapper but fixed with MSI. So the image in MSI screens should be vertical. Mixxx logo doesn't fit unless you turn it 90° anti-clockwise.

Be-ing commented:
> the image in MSI screens should be vertical

Is transparency supported? If so, you could hack a vertical image by adding a transparent background below.

sblaisot committed:
(wix) fix bootstrapper welcome screen text cropping

sblaisot committed:
(wix) UI polishing, remove screenshot, black theme for bundle

sblaisot committed:
Merge branch 'wix-bundle' of https://github.com/sblaisot/mixxx into wix-bundle

sblaisot committed:
(wix) use fixed ProductID when building bundle to fix uninstallation

sblaisot committed:
(wix) fix registry key creation on 32bits windows

sblaisot commented:
ok, I reworked the bootsrapper user interface. It now looks like this:
![image](https://user-images.githubusercontent.com/2768193/28292312-c763cc6a-6b4e-11e7-8a27-ff5b227e8638.png)
No more "sidebar" and a big Mixxx logo on top replacing the window Title.

I also reworked embedded MSI interface to remove Mixxx screenshot. Here is the result:
![image](https://user-images.githubusercontent.com/2768193/28292397-0b6a373c-6b4f-11e7-8221-9a1dc919dace.png)

Uninstallation should be fixed now.
It appears to be a bug somewhere but I couldn't find where. 
The only reference I found about this was that [StackOverflow question](https://stackoverflow.com/questions/31433770/burn-bootstrapper-does-not-uninstall-chained-msi)
As a workaroud, I used a fixed ProductID, generated at each compilation by scons.

So we are now waiting appveyor to product an artifact and we will be able to test again.

Be-ing commented:
>  Black background, no more "sidebar" and a big Mixxx logo on top replacing the window Title.

I like how that looks, but I think it would be weird to have the color scheme totally change part way through the installation process.

sblaisot commented:
I can not, at the moment, move the feature selection up at the bundle level. So for now, I have no other choice than to display the embedded MSI User Interface.
So practically, you have two installer windows, one on top of the other (the bundle interface first, then the embedded MSI interface).
This explains the two look and feel (but as it's in another window, it doesn't look like a look and feel change and it's a way to distinguish between them).
This is not ideal, but I have no other option right now.

sblaisot commented:
@daschuer @esbrandt can you test the last version please ?

sblaisot committed:
Merge branch 'master' into wix-bundle

esbrandt commented:
Latest ``mixxx-2.1.0-alpha-pre-master-PR1306-git6324-release-x64.exe`` works fine in Win 10 VM. Install, modify, uninstall - all fine. Thanks.
 
re: Color scheme.
The font color is white with a shadow, and i found it looks kinda jagged on a black background. Can´t we go the standard route, black typeface on white background?

![4](https://user-images.githubusercontent.com/4525897/28494091-8b3daf38-6f24-11e7-8207-acd7b5d13a1e.PNG)

sblaisot commented:
I agree the title looks ugly with its shadow.
I can change back to white background. I can also switch the title to plai, white instead of grey+shadow.

sblaisot commented:
ok, I finally managed to remove the shadow on the title in the sucess window.
here are the comparison between black and white background.
We need to chosse between the two.
Waiting to hear from the UX/UI lead designer ;)

![mixxx setup done - test4](https://user-images.githubusercontent.com/2768193/28499826-190dcf38-6fbf-11e7-8d4e-e29fb545fb03.jpg)![mixxx setup done - test3](https://user-images.githubusercontent.com/2768193/28499823-0eec2dce-6fbf-11e7-845e-06082b14409e.jpg)

Be-ing commented:
I tested this on my friend's computer. When I first tried to install, I had to click through several warnings about unsigned packages. I presume that's because the build was from AppVeyor? Does our build server sign the packages?

After that, I got an error saying it could not install because an old version of Mixxx was installed and it said to uninstall that first, but it gave no instructions on how to uninstall an old version of Mixxx. This will be a stumbling block for lots of users upgrading. Are we able to automatically run the uninstaller for the old version? Or make a link in the warning dialog to start the uninstaller? Or at least a link to the part of the Windows Control Panel for uninstalling programs? Failing those, we should describe how to get to the Control Panel. This may be different on different versions of Windows. If so, if the installer can detect which version of Windows is being used, it should present the appropriate instructions for that version of Windows. Otherwise, it should show instructions for every supported version of Windows.

Why are the debugging files and desktop shortcut not installed by default? Do we really need to present a screen with options to omit the translations and user manual?

Uninstallation worked for the most part. There were a few old files left over from 2.0 or 2.0 beta, namely controllers/DN-HS5500_MIDI_1.midi.xml and empty "promo" folder. When I manually deleted those, reinstalled the new version, and uninstalled it, all files were deleted.

sblaisot commented:
* Package signing
no, we don't sign our windows packages, nor linux packages. We have a wishlist bug for that: https://bugs.launchpad.net/mixxx/+bug/1517823
The main problems are that:
1) someone need to pay for a certificate (there are no free (as in free beer) code signing cert providers)
2) certificate lifetime are short (1 year or 2 years max), so we need to pay each year or so
3) build process needs to be reworked to include package signing

* need to manually uninstall previous versions
We can NOT automate previous version uninstallation because #836 has been merged AFTER 2.0 release. The plan was to have it shipped with 2.0.1 but this version never came up.
We can add instructions on how to remove but I'm afraid we can not launch uninstaller from within the MSI
Don't forget MSIs are NO CODE, but only a bunch of description tables and some packages (cab file), so we are very limited and can't do everything we want/need.

* Feature selection
debug files (pdb) cost a lot on hard disk and lots of people have no usage for that.
and I personally HATE programs that put a link on my desktop without asking the permission, so it's a nogo from me to build an installer that do this.

* files left by NSIS installer: That's one of the main reason to switch to wix :)
https://bugs.launchpad.net/mixxx/+bug/1457618

sblaisot commented:
btw, thanks a lot for testing and your feedback

Be-ing commented:
:/ This upgrade situation is not good. I guess we need very detailed step-by-step instructions verified for every supported version of Windows.

> debug files (pdb) cost a lot on hard disk and lots of people have no usage for that.

Won't it be hard to get useful bug reports if users don't have these installed? Also the installer says they take up 0 kb.

Be-ing commented:
Uninstallation instructions:
  * [Windows 10](https://www.laptopmag.com/articles/uninstall-programs-windows-10)
  * [Windows 8](http://www.dummies.com/computers/operating-systems/windows-8/add-or-remove-programs-in-windows-8/)
  * [Windows 7](https://support.hp.com/th-en/document/c01835723)
  * [Windows Vista](https://support.hp.com/us-en/document/c01571960)
  * [Windows XP](https://support.microsoft.com/en-us/help/307895/how-to-change-or-remove-a-program-in-windows-xp)

sblaisot commented:
we can shorten uninstall instructions to %PROGRAMFILES%\Mixxx\UninstallMixxx.exe ;)

sblaisot commented:
PDB files cost about 100MB when you select it. It costs 0KB while unselected (the default).

Be-ing commented:
> we can shorten uninstall instructions to %PROGRAMFILES%\Mixxx\UninstallMixxx.exe ;)

Can we hack that into a local file link that Windows Explorer would open and start running the uninstaller?

Be-ing commented:
It looks like it is possible to run a command from Wix:
https://stackoverflow.com/questions/1819686/wix-open-web-page-when-uninstall-completes?rq=1
http://wixtoolset.org/documentation/manual/v3/wixdev/extensions/authoring_custom_actions.html

sblaisot commented:
Properly handling upgrade path is beyond the scope of this PR.
Please keep this one focused on MSVC reditributable being correctly installed and I will work on that after this one has been merged.

Yes, it is possible to launch an executable from an MSI. This is exactly what we did before for MSVC redistributable and that didn't work well.
Now, with the bootstrapper, we have more flexibility.
I would be very happy to share my headache on windows installer with someone if there is a volunteer to take care of this. a c-sharp dev could create a custom bootstrapper. Unfortunately, this is not one of my skills and not something I'm interested in.

Be-ing commented:
Okay, then I think this should be merged as-is. It will already be helpful for users testing master builds.

> Yes, it is possible to launch an executable from an MSI. This is exactly what we did before for MSVC redistributable and that didn't work well.

But that was an executable in the installation package, right? I am referring to running the Mixxx uninstaller executable that the user already has on their computer.

sblaisot committed:
(wix) fix bundle window title unwanted shadow

mevsme commented:
Why can not we roll the new version over the old one in Windows?

sblaisot commented:
what do you mean by "roll the new version over the old one" ?

if you think of simply put the new files on top of the older ones (overwriting them), you will have two uninstall entries in "Add/remove programs" and thinks like that and you will keep old files unmaintained with probable conflicts (like old skin files and so on).

this is not a good idea at all.

sblaisot commented:
Hi,

This is already an improvement over existing installer. This PR corrects the MSVC redistributable installation.

@daschuer can we merge this please ?
I will work on remaining issues like upgrade path in another PR.

thanks,

sb

sblaisot committed:
(wix) improve productID autogeneration

sblaisot commented:
can we please merge this so that I can work on next steps ?

TIA

sb

daschuer commented:
Oh Sorry, I must have missed this. Thank you for all the work.

daschuer commented:
It looks like the CI issues are not related to this PR, so I will merge now.
